Firebase.h

```diff
@@ -24,50 +24,59 @@
 #include <ESP8266WiFi.h>
 #include <WiFiClientSecure.h>
 #include <ESP8266HTTPClient.h>
+#include <ArduinoJson.h>
 
-// FirebaseError represents a Firebase API error with a code and a
-// message.
-class FirebaseError {
+// FirebaseError is an error for a given firebase API call.
+struct FirebaseError {
+  FirebaseError() {}
+  FirebaseError(int c, const String& m) : code(c), message(m) {}  
+  FirebaseError(int c, const char* m) : code(c), message(m) {}
+  int code = 0;
+  String message = "";
+  operator bool() const {
+    return code < 0;
+  }
+};
+
+// FirebaseObject is a payload or a result for a given firebase API call.
+class FirebaseObject {
  public:
-  operator bool() const { return _code < 0; }
-  int code() const { return _code; }
-  const String& message() const { return _message; }
-  void reset() { set(0, ""); }
-  void set(int code, const String& message) {
-    _code = code;
-    _message = message;
+  FirebaseObject() : json(_buf.createObject()) {
+  }
+  FirebaseObject(const String& data) : _data(data), json(_buf.parseObject((char*)_data.c_str())) {
+    if (!json.success()) {
+      error = FirebaseError{-1, "error parsing json"};
+    }
+  }
+  FirebaseObject(const FirebaseError& err) : error{err}, json(_buf.createObject()) {
   }
  private:
-  int _code = 0;
-  String _message = "";
+  StaticJsonBuffer<200> _buf;
+  String _data;
+ public:
+  FirebaseError error;
+  JsonObject& json;
 };
 
-// Firebase is the connection to firebase.
+// Firebase is a client for a given firebase host.
 class Firebase {
  public:
   Firebase(const String& host);
   Firebase& auth(const String& auth);
-  const FirebaseError& error() const {
-    return _error;
-  }
-  String get(const String& path);
-  String push(const String& path, const String& value);
+  FirebaseObject get(const String& path);
+  FirebaseObject push(const String& path, const FirebaseObject& value);
   bool connected();
```

Comment:
I won't force the issue but I still feel like the concept of being connected or available is only relevant to one usecase (the stream call) so it seems like this should all be separated from the Firebase object to keep it as simple as possible to understand for the other uses.